crates/cargo-contract/src/cmd/extrinsics/balance.rs

```diff
@@ -0,0 +1,274 @@
+// Copyright 2018-2022 Parity Technologies (UK) Ltd.
+// This file is part of cargo-contract.
+//
+// cargo-contract is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// cargo-contract is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with cargo-contract.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::fmt::Display;
+
+use serde_json::json;
+
+use super::{
+    Balance,
+    Client,
+};
+
+use anyhow::{
+    Context,
+    Ok,
+    Result,
+};
+
+/// Represents different formats of a balance
+#[derive(Debug, Clone)]
+pub enum BalanceVariant {
+    /// Default format: no symbol, no denomination
+    Default(Balance),
+    /// Denominated format: symbol and denomination are present
+    Denominated(String),
+}
+
+#[derive(Debug, Clone)]
+pub struct TokenMetadata {
+    /// Number of denomination used for denomination
+    pub denomination: u128,
+    /// Token symbol
+    pub symbol: String,
+}
+
+impl TokenMetadata {
+    /// Query [TokenMetadata] through the node's RPC
+    pub async fn query(client: &Client) -> Result<Self> {
+        let sys_props = client.rpc().system_properties().await?;
+
+        let default_decimals = json!(12);
+        let default_units = json!("UNIT");
+        let decimals = sys_props
+            .get("tokenDecimal")
+            .unwrap_or(&default_decimals)
+            .as_u64()
+            .context("error converting decimal to u64")?;
+        let symbol = sys_props
+            .get("tokenSymbol")
+            .unwrap_or(&default_units)
+            .as_str()
+            .context("error converting symbol to string")?;
+        let denomination: u128 = format!("1{}", "0".repeat(decimals as usize)).parse()?;
+        Ok(Self {
+            denomination,
+            symbol: symbol.to_string(),
+        })
+    }
+}
+
+impl BalanceVariant {
+    /// Converts BalanceVariant into Balance.
+    ///
+    /// Throws Error if [BalanceVariant::Denominated(String)] is in an incorrect format.
+    pub fn denominate_balance(&self, token_metadata: &TokenMetadata) -> Result<Balance> {
+        match self {
+            BalanceVariant::Default(balance) => Ok(*balance),
+            BalanceVariant::Denominated(input) => {
+                if let Some(balance_str) =
+                    input.strip_suffix(&format!("k{}", token_metadata.symbol))
+                {
+                    let balance: f64 = balance_str.parse()?;
```

Comment:
We should avoid using floats for dealing with currency amounts, see https://stackoverflow.com/questions/3730019/why-not-use-double-or-float-to-represent-currency.

I have used https://docs.rs/rust_decimal/latest/rust_decimal/ before, that is worth a look.

Reply:
Thanks! I used `Decimal` for `denominate_balance()`. I thought of using it also in `from<from<T: Into<u128>>>()`, but `Decimal::new()` only takes `i64` and `u32` according to [docs](https://docs.rs/rust_decimal/latest/rust_decimal/struct.Decimal.html#method.new), so that would imply using `TryInto` which would add unnecessary type conversions and `Result` as a return type

Reply:
One issue is that there could be some loss of precision with `rust_decimal`, it seems the [max value](https://github.com/paupino/rust-decimal/blob/master/src/decimal.rs#L201) is 96 bits https://github.com/paupino/rust-decimal/blob/master/src/decimal.rs#L464.

Thinking out loud we need:
1. Carefully protect against loss of precision when performing conversions in `denominate_balance`, although this might be accounted for already in overflow panics. If this is the case these should be tested and handled gracefully, because we would be relying on the underlying library panicking on overflow rather than, say wrapping around.
2. Research alternative libraries which would allow storing up to `u128` precision decimals?
3. Define a custom built `Decimal` type e.g. 
```
pub struct Decimal {
    value: u128,
    scale: i8,
}
```
And then provide some specialised methods to scale by exponents, which is the main thing we need.